src/python/BuildingControlsSimulator/Simulator/Simulation.py

```diff
@@ -28,136 +29,174 @@ class Simulation:
+        """Main co-simulation loop"""
+        logger.info("Initializing co-simulation models")
+        self.initialize(data_period=data_period)
 
-        logger.info("Running co-simulation ...")
-        output = []
-        t_ctrl = self.building_model.init_temperature
-        air_temp_output_idx = self.get_air_temp_output_idx()
+        sim_data_channel_idx_offset = self.data_client.hvac.data[
+            self.data_client.hvac.data[STATES.DATE_TIME] == data_period[0]
+        ].index.values[0]
 
-        for t_step_seconds in range(
+        # pre-allocate output arrays
+        sim_time = np.arange(
             self.start_time_seconds,
             self.final_time_seconds,
             self.step_size_seconds,
-        ):
-            controller_output = self.controller_model.do_step(t_ctrl)
+            dtype="int64",
+        )
 
-            self.building_model.actuate_HVAC_equipment(
-                self.controller_model.HVAC_mode
-            )
-            step_status = self.building_model.fmu.do_step(
-                current_t=t_step_seconds,
-                step_size=self.step_size_seconds,
-                new_step=True,
-            )
-            # save data as row
-            building_model_output = np.array(
-                [
-                    self.building_model.fmu.get(k)[0]
-                    for k in self.building_model_output_keys
-                ]
+        logger.info(
+            f"Running co-simulation from {data_period[0]} to {data_period[1]}"
+        )
+        for i in range(0, len(sim_time)):
+            self.controller_model.do_step(
+                t_start=sim_time[i],
+                t_step=self.step_size_seconds,
+                step_hvac_input=self.data_client.hvac.data.iloc[
+                    sim_data_channel_idx_offset + i
+                ],
+                step_sensor_input=self.building_model.step_output,
+                step_weather_input=self.data_client.weather.data.iloc[
+                    sim_data_channel_idx_offset + i
+                ],
             )
-
-            t_ctrl = self.calc_T_control(
-                building_model_output, air_temp_output_idx
+            self.building_model.do_step(
+                t_start=sim_time[i],
+                t_step=self.step_size_seconds,
+                step_control_input=self.controller_model.step_output,
+                step_sensor_input=self.data_client.sensors.data.iloc[
+                    sim_data_channel_idx_offset + i
+                ],
+                step_weather_input=self.data_client.weather.data.iloc[
+                    sim_data_channel_idx_offset + i
+                ],
             )
 
-            step_output = (
-                [t_step_seconds, step_status, t_ctrl]
-                + controller_output
-                + list(building_model_output)
-            )
-            output.append(step_output)
-            # TODO add multizone support
+        logger.info("Finished co-simulation")
```

Comment:
Might be useful here to measure and log the time needed to perform the simulation.

Reply:
good idea, adding this.